### PR TITLE
Optional intermediary bridge netdev

### LINKED ISSUE #1577: Connections going via FDP don't appear in conntrack
Which kinda breaks scope.

Repro:
```
docker run -d --name nginx nginx
docker run -d --name client alpine /bin/sh -c "while true; do \
	wget http://nginx.weave.local:80/ -O - >/dev/null || true; \
	sleep 1; \
done"
sudo conntrack -E -p tcp
```

Pre-1.2 (or with WEAVE_NO_FASTDP=true) you see conntrack events for the connections coming and going.

### PR DESCRIPTION
**Pull request by awh**
Modify the weave script to interpose an intermediary bridge netdev between application containers and the OVS datapath. Fixes #1577.

### CODE REVIEW

**Comment by awh** [comment]
Hmmm need to think about plugin...

**Comment by awh** [comment]
Tested the repro in #1577 against Scope 0.10.0 + this branch; as you can see short lived connections to port 80 are visible:
fastdatapath-conntrack
(note the spurious connection to port 4040 appears to be an instance of weaveworks/scope#520)

File: weave [hunk 1]
```diff
@@ -495,6 +509,39 @@ init_bridge_fastdp() {
...
+init_bridge_bridged_fastdp() {
+    # Initialise the datapath as normal. NB sets MTU for use below
+    init_bridge_fastdp
+
+    # Create linking veth pair
+    ip link del $FASTDP_BRIDGE_IFNAME >/dev/null 2>&1 || true
+    ip link del $BRIDGE_IFNAME >/dev/null 2>&1 || true
+    ip link add name $FASTDP_BRIDGE_IFNAME mtu $MTU type veth peer name $BRIDGE_IFNAME mtu $MTU || return 1
+
+    # Create intermediary bridge netdev
```
Comment (bboreham) [inline weave]: Following 10 lines seem to duplicate `init_bridge_bridge`

File: weave [hunk 2]
```diff
@@ -495,6 +509,39 @@ init_bridge_fastdp() {
...
+    # Set MTU
+    ip link add name v${CONTAINER_IFNAME}du mtu $MTU type dummy
+    ip link set dev v${CONTAINER_IFNAME}du master $FASTDP_BRIDGE
+    ip link del dev v${CONTAINER_IFNAME}du
+
+    # Link intermediary bridge and datapath
+    if ! ip link set $FASTDP_BRIDGE_IFNAME up ||
+       ! ip link set $BRIDGE_IFNAME up ||
+       ! add_iface_fastdp $BRIDGE_IFNAME || ! ip link set $FASTDP_BRIDGE_IFNAME master $FASTDP_BRIDGE ; then
+        ip link del $BRIDGE_IFNAME >/dev/null 2>&1 || true
```
Comment (bboreham) [inline weave]: could comment here "clean up if something went wrong"

File: weave [hunk 3]
```diff
@@ -402,20 +407,56 @@ random_mac() {
...
-    if [ ! -d /sys/class/net/$BRIDGE ] ; then
-        if [ -n "$WEAVE_NO_FASTDP" ] ; then
-            BRIDGE_TYPE=bridge
-        elif docker run --rm --privileged --net=host $IMAGE $COVERAGE_ARGS --create-datapath --datapath=$BRIDGE ; then
-            BRIDGE_TYPE=fastdp
-        elif [ $? = 17 ] ; then
-            # Exit status of 17 means the kernel doesn't have ODP
+    if [ -d /sys/class/net/$DATAPATH ] ; then
+        # Looks like we're using bridged_fastdp so far
+        if [ ! -d /sys/class/net/$DATAPATH/bridge -d /sys/class/net/$BRIDGE/bridge ] ; then
```
Comment (bboreham) [inline weave]: is two `-d`s in a row a typo?

Comment (awh) [inline weave]: Hah yes it was. I'm still wrangling atm let me assign the PR to me as it is under active development atm

File: weave [hunk 4]
```diff
@@ -108,6 +108,7 @@ exec_remote() {
         -e WEAVE_CONTAINER_NAME \
         -e WEAVE_MTU \
         -e WEAVE_NO_FASTDP \
+        -e WEAVE_NO_BRIDGED_FASTDP \
```
Comment (bboreham) [inline weave]: Should `WEAVE_NO_BRIDGED_FASTDP` be documented as `WEAVE_NO_FASTDP` is?

Comment (rade) [inline weave]: no. And we should remove the (docs for the) latter soon.

File: weave [hunk 5]
```diff
@@ -1689,6 +1781,9 @@ EOF
         fi
         create_bridge --without-ethtool
         ;;
+    bridge-type)
```
Comment (bboreham) [inline weave]: Is this used anywhere?  Just for debugging?

**Comment by bboreham** [comment]
Very nice.

File: weave (hunk 4 above)
Comment (awh) [inline weave]: The main reason I included this was to lay groundwork for #1681, but it was also useful in comparative performance testing.

File: weave (hunk 5 above)
Comment (awh) [inline weave]: Yes, purely for debugging and manual testing. Shall I remove it?

**Pull request merged**